src/terminal.rs

```diff
@@ -1,19 +1,38 @@
 #[cfg(unix)]
 pub(crate) fn acquire_terminal(interactive: bool) {
     use is_terminal::IsTerminal;
-    use nix::sys::signal::{signal, SigHandler, Signal};
+    use nix::{
+        errno::Errno,
+        sys::signal::{signal, SigHandler, Signal},
+        unistd,
+    };
 
-    if !std::io::stdin().is_terminal() {
-        return;
-    }
+    if interactive && std::io::stdin().is_terminal() {
+        // see also: https://www.gnu.org/software/libc/manual/html_node/Initializing-the-Shell.html
 
-    take_control(interactive);
+        take_control();
+
+        unsafe {
+            // SIGINT and SIGQUIT have special handling above
+            signal(Signal::SIGTSTP, SigHandler::SigIgn).expect("signal ignore");
+            signal(Signal::SIGTTIN, SigHandler::SigIgn).expect("signal ignore");
+            signal(Signal::SIGTTOU, SigHandler::SigIgn).expect("signal ignore");
+        }
 
-    unsafe {
-        // SIGINT and SIGQUIT have special handling above
-        signal(Signal::SIGTSTP, SigHandler::SigIgn).expect("signal ignore");
-        signal(Signal::SIGTTIN, SigHandler::SigIgn).expect("signal ignore");
-        signal(Signal::SIGTTOU, SigHandler::SigIgn).expect("signal ignore");
+        // Put ourselves in our own process group and take control of terminal, if not already
+        let shell_pgid = unistd::getpid();
+        match unistd::setpgid(shell_pgid, shell_pgid) {
+            // setpgid returns EPERM if we are the session leader (e.g., as a login shell).
+            // The other cases that return EPERM cannot happen, since we gave our own pid.
+            // See: setpgid(2)
+            // Therefore, it is safe to ignore EPERM.
+            Ok(()) | Err(Errno::EPERM) => (),
+            Err(_) => {
+                eprintln!("ERROR: failed to put nushell in its own process group");
+                std::process::exit(1);
```

Comment:
Does this really need to exit the entire shell? Same with the exit(1) below.

Reply:
This is a good question. The GNU reference and `fish` both error in this case, so I was initially copying them. Looking in to it more, it is probably best to ensure that `setpgid` succeeds. This ensures nushell is in its own process group which helps with signal handling and prevents orphaning.
See: https://github.com/fish-shell/fish-shell/commit/a2e486966acbb15b1c666ab22e81f42919b67353 and https://github.com/fish-shell/fish-shell/issues/7060.

Consider if `setpgid` fails and we continue, then every nushell instance may or may not be in its own process group. If users or other programs expect one or the other, then they may be in for a surprise! I don't think `setpgid` should be able to error in this case anyways, since we are passing our own pid to it.
See: https://man7.org/linux/man-pages/man2/setpgid.2.html#ERRORS

For the "exit(1) below", are you referring to the three calls in `take_control`? Or one of them in particular? I think we need these as well, since they handle the case where we (probably) have no terminal but are supposed to be running in interactive mode.

Reply:
I'm good with deferring to whatever your opinion is since you've done your research. I just think exit is kind of final. In some situations, that is warranted. In others, it may not be. I just don't know. So, let's run with your choice and see how it goes. Do you have anything else we need to do here or are you ready to land?

Reply:
One limitation of `std::process::exit()` is that it hands everything back to the OS so no unwinding/`Drop` is happening. This is fine with everything that is just memory or file descriptors but could be a problem if we use `Drop` to clean up. If I read it correctly we don't launch anything too bad (e.g. setting terminal raw modes only happens in `run_repl()` afterwards) so it could be fine.

Reply:
Yep, the exit is definitely a valid concern! In this case, I think it is warranted. I'm slightly less sure about the other three, but they were already there and I simply chose to leave them mostly as is. Otherwise, this should be good go. Thanks!